File: Python/textbook_track/chapter02/ba2a.py
# Implement MotifEnumeration
import sys

NUCLEOTIDES = ('A', 'C', 'G', 'T')


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def calc_hamming_distance(s1, s2):
    distance = 0
    for c1, c2 in zip(s1, s2):
        if c1 != c2:
            distance += 1
    return distance


def calc_immediate_neighbours(pattern):
    neighbours = set()
    for ix, letter in enumerate(pattern):
        for nucleotide in NUCLEOTIDES:
            if nucleotide != letter:
                pattern_list = list(pattern)
                pattern_list[ix] = nucleotide
                neighbours.add(''.join(pattern_list))
    return neighbours


def generate_neighbourhood(text, d):
    neighbourhood = {text}
    for _ in range(d):
        neighbours_of_neighbours = set()
        for pattern in neighbourhood:
            immediate_neighbours = calc_immediate_neighbours(pattern)
            neighbours_of_neighbours.update(immediate_neighbours)
        neighbourhood.update(neighbours_of_neighbours)
    return neighbourhood


def does_k_mer_occurs_approximately_in_text(text, k_mer, k, d):
    for ix in range(len(text)-k+1):
        substring = text[ix:ix+k]
        if calc_hamming_distance(substring, k_mer) <= d:
            return True
    return False


def motif_enumeration(texts, k, d):
    def all_texts_approximately_contain_pattern(k_mer):
        return all(does_k_mer_occurs_approximately_in_text(text, k_mer, k, d) for text in texts)

    motifs = set()
    for text in texts:
        for ix in range(len(text)-k+1):
            pattern = text[ix:ix+k]
            neighbours = generate_neighbourhood(pattern, d)
            for neighbour in neighbours:
                if all_texts_approximately_contain_pattern(neighbour):
                    motifs.add(neighbour)
    return motifs


def main():
    data = sys.stdin.read().splitlines()
    k, d = convert_to_intlist(data[0])
    result = motif_enumeration(data[1:], k, d)
    print(' '.join(result))


if __name__ == '__main__':
    main()

File: Julia/textbook_track/chapter02/ba2a.jl
# Implement MotifEnumeration

const NUCLEOTIDES = ('A', 'C', 'G', 'T')


convert_to_intlist(line) = map(x -> parse(Int, x), split(line))


calc_hamming_distance(s1, s2) = count(((c1, c2),) -> c1 != c2, zip(s1, s2))


function calc_immediate_neighbours(pattern)
    neighbours = Set{String}()
    for (ix, letter) in enumerate(pattern)
        for nucleotide in NUCLEOTIDES
            if nucleotide != letter
                pattern_array = collect(pattern)
                pattern_array[ix] = nucleotide
                push!(neighbours, join(pattern_array))
            end
        end
    end
    neighbours
end


function generate_neighbourhood(text, d)
    neighbourhood = Set{String}([text])
    for _ in 1:d
        neighbours_of_neighbours = Set{String}()
        for pattern in neighbourhood
            immediate_neighbours = calc_immediate_neighbours(pattern)
            union!(neighbours_of_neighbours, immediate_neighbours)
        end
        union!(neighbourhood, neighbours_of_neighbours)
    end
    neighbourhood
end


function does_k_mer_occurs_approximately_in_text(text, kmer, k, d)
    for ix in 1:(length(text)-k+1)
        substring = text[ix:ix+k-1]
        if calc_hamming_distance(substring, kmer) <= d
            return true
        end
    end
    false
end


function motif_enumeration(texts, k, d)
    all_texts_approximately_contain_pattern(kmer) =
        all(does_k_mer_occurs_approximately_in_text(text, kmer, k, d) for text in texts)

    motifs = Set{String}()
    for text in texts
        for ix in 1:(length(text)-k+1)
            pattern = text[ix:ix+k-1]
            neighbours = generate_neighbourhood(pattern, d)
            for neighbour in neighbours
                if all_texts_approximately_contain_pattern(neighbour)
                    push!(motifs, neighbour)
                end
            end
        end
    end
    motifs
end


function main()
    k, d = convert_to_intlist(readline())
    texts = readlines()
    result = motif_enumeration(texts, k, d)
    println(join(result, " "))
end


main()
